### ISSUE TITLE
native_op.ctc_loss: non-zero gradient (= softmax) for sequences with no valid alignment

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1835]
`returnn.tf.native_op.ctc_loss` returns loss 0 for a sequence with no valid alignment, but its
error signal for that sequence is **exactly `softmax(logits)`** rather than 0. The sequence then
contributes a full-magnitude, wrong-direction gradient to training, silently: nothing is `inf` or
`nan`, and the loss looks harmless.

### Measurements

Vocab 6, blank = 5, single sequence, float32.

| case | loss | mean abs grad | max abs grad | grad vs `softmax(logits)` | grad sum per frame |
|---|---|---|---|---|---|
| feasible, T=8, S=3 | 11.90169239 | 0.199641 | 0.693012 | 9.6e-01 | ~1e-7 |
| targets > input, T=2, S=3 | 0.00000000 | 0.166667 | 0.590199 | **1.5e-08** | **1.0** |
| repeated labels, T=2, S=2 `[1,1]` | 0.00000000 | 0.166667 | 0.590199 | **1.5e-08** | **1.0** |

The per-frame sum is the clearest symptom: a correct CTC gradient is `p - posterior`, both
normalised, so it sums to 0 over the vocab axis. Here it sums to exactly 1.0.

Deterministic, and identical on CPU and GPU (H100): repeated runs give bit-identical values, and
`grad == softmax(logits)` to float32 precision.

### Cause

In `ctc_loss`:

```python
fwdbwd, obs_scores = fast_baum_welch(am_scores=-log_sm, float_idx=seq_mask, edges=..., ...)
loss = obs_scores[0]
bw = tf.exp(-fwdbwd)
if grad_wrt_softmax_in:
    grad_x = tf.exp(log_sm) - bw
```

With no valid path, `fast_baum_welch` gives `fwdbwd = inf`, so `bw = 0` and `grad_x` degenerates
to `exp(log_sm)` -- the posterior term is simply absent. `obs_scores` comes back as 0 rather than
`inf` for those sequences, which is why the loss looks fine.

### Why it matters

Real data contains such sequences: targets that do not fit the subsampled input, or repeated
adjacent labels with too few frames to place the required blank. On a production Loquacious
conformer AED+CTC setup they occur often enough to have crashed `tf.nn.ctc_loss` (which returns
`inf` for the same inputs) within the first epoch.

Descending on `p` alone flattens the distribution toward uniform, i.e. it actively raises entropy
on exactly the frames that carry no supervision, at about the same gradient magnitude as a real
sequence (mean 0.167 vs 0.200 above).

For comparison, `TorchBackend.ctc_loss` passes `zero_infinity=True` to
`torch.nn.functional.ctc_loss`, which contributes **0 loss and 0 gradient**. So the two backends
currently disagree on the same data.

### Affected

- `CtcLossLayer` and `CtcLoss(use_native=True)`, i.e. net-dict CTC setups.
- The RF TF backend, where `TFBackend.ctc_loss` now defaults to the native op (#1834). Worked
  around there at the backend level, since that is new code with no existing users: it masks on
  `input_len >= target_len + number of adjacent repeated labels`, which zeroes the gradient too
  (`tf.where`, and the error signal is finite, so no `0 * nan`).
- The PyTorch backend only if `use_native_op=True` is passed explicitly; its default is False.

Only `ctc_loss` was measured, not `ctc_loss_viterbi`.

### Proposed fix

Do the same masking inside `native_op.ctc_loss`, so every caller gets it.

`is_finite` cannot be used to detect these: the loss is already 0, indistinguishable from a
genuinely small one on a confident model. Feasibility from the lengths is exact and cheap:
`input_len >= target_len + adjacent_repeats` (the repeats term only with `label_loop`).
Alternatively `fast_baum_welch` could report no-path explicitly instead of returning 0.

**This changes gradients for existing trainings** whose data contains such sequences, so it
presumably wants a behaviour version rather than a silent change -- flagging rather than
deciding.

### Environment

TF 2.20.0, CUDA 12.5.1 build, Python 3.12.3, NVIDIA H100 and CPU, RHEL 9.

**Comment by albertz** [comment]
Viterbi numbers (`ctc_loss_viterbi`, via `fast_viterbi` + CE on the alignment), analogous to the `ctc_loss` measurement above. T=10, B=3, D=5, blank=4; seq 0: 3 labels, valid; seq 1: 12 labels in 10 frames, impossible; seq 2: 6 labels with adjacent repeats (min 9 frames), valid. TF 2.20, H100, native op:

|  | loss per seq | sum(abs(grad)) per seq |
|---|---|---|
| `zero_infinity=False` | [11.34, **inf**, 17.82] | [12.46, **15.25**, 15.48] |
| `zero_infinity=True` | [11.34, **0**, 17.82] | [12.46, **0**, 15.48] |

So without the option, an impossible sequence contributes an `inf` loss (poisons the whole batch score) and a gradient of the same magnitude as a healthy sequence: `fast_viterbi` fills its alignment with the mask index, and the cross entropy then trains toward that filler label. The behavior is deterministic (identical across repeated runs). With `zero_infinity=True`, the sequence contributes exactly 0 loss and 0 gradient, and the valid sequences are bit-identical in both modes.
